docs/rfd.md

```diff
@@ -0,0 +1,478 @@
+
+2. Streaming output:
+
+```go
+stream, err := client.Output(cmd.Context(), &jobv1.OutputRequest{
+  Id:     jobId,
+  Follow: true,
+})
+if err != nil {
+  return err
+}
+for {
+  resp, err := stream.Recv()
+  if err != nil {
+    return err
+  }
+  fmt.Fprint(os.Stdout, resp.GetOutput())
+}
+```
+
+3. Stopping the job:
+
+```go
+_, err := client.Stop(ctx, jobId)
+```
+
+#### CLI
+
+The CLI is the primary means of interacting with the Job Server. It will have a familiar interface to other CLI tools like `docker` or `kubectl`.
+The following examples demonstrate how the CLI will be used:
+
+```bash
+# starting the job server
+$ sudo jobserver serve --listen-address=127.0.0.1:9090 --rbac=rbac-config.yaml \
+  --cacert=ca.crt --cert=tls.crt --key=tls.key
+
+# to use the client, flags to configure cert and key paths, and the server address, are required.
+# these are omitted in the remaining examples for brevity.
+$ jobserver --address=127.0.0.1:9090 --cacert=ca.crt --cert=tls.crt --key=tls.key ...
+
+# starting a job:
+# for commands that don't require flag args, they can be passed as-is:
+$ jobserver run kubectl logs pod/example
+<job-id> # job id is printed
+
+# for commands that do require flag args, the expected '--' delimiter can be used:
+$ jobserver run -- kubectl logs --follow pod/example
+<job-id>
+
+# to start a job with resource limits:
+$ jobserver run \
+  --cpus=100m \ # 100 milli-cores; can also write e.g. '--cpus=2' for 2 cores (no float values)
+  --memory=1Gi --memory-soft-limit=512Mi \ # the familiar suffixes can be used for memory limits
+  --device-read-bps=8:16=200 \ # format is major:minor=value (bytes per second)
+  --device-write-bps=8:16=200,8:0=50 \ # multiple devices can be specified by separating with commas
+  --device-read-iops=... --device-write-iops=... \ # same format, but units are in iops
+  -- go build -o bin/jobserver ./cmd/jobserver
+
+# stopping a job:
+$ jobserver stop <job-id>
+<job-id>
+
+# getting the status of a job:
+$ jobserver status <job-id>
+state: running
+pid: 1234
+start time: 2006-01-02T15:04:05Z07:00
+
+# getting the output of a job:
+$ jobserver logs [--follow] <job-id>
+<output>
+
+# getting a list of all job ids:
+$ jobserver list
+JOB ID      COMMAND       CREATED          STATUS
+<id-1>      kubectl       2006-01-02...    Running
+<id-2>      go            2006-01-02...    Completed
+```
+
+### Implementation Details
+
+#### Streaming Output
+
+Output streaming will be implemented as follows:
+
+1. When a job is started, we will pipe the combined stdout and stderr of its process to a simple in-memory buffer.
+2. When a client initiates a stream by calling `Output()`, we will write the current contents of the buffer to the stream in 4MB chunks. Then, if the client requested to follow the output, the server will keep the stream open, and any subsequent reads from the process's output will be duplicated to all clients that are following the stream, in addition to the server's own buffer.
```

Comment:
4MiB is the default message size limit but it's far larger than the recommended size for messages in a stream; "oral tradition" puts that at 16-64KiB (https://github.com/grpc/grpc.github.io/issues/371) apparently.

Reply:
Message size limits are tricky. The way I understand it is that regardless of grpc message size, the underlying http/2 transport will chunk it into frames (16KiB is the default frame size and as far as I know grpc doesn't change it), but grpc itself adjusts the transport's flow control windows (unless you change the default window size manually)

From my own testing, using larger messages results in significantly higher throughput, ostensibly because reading and writing larger messages triggers grpc to increase the size of the flow control windows, which results in fewer window update frames required per rpc. 

This is all undocumented of course, and I've never been able to find a comprehensive guide on this topic, but that's what I have gathered by looking through the grpc and net/http2 sources.

Reply:
That's pretty neat, although 4MiB still feels uncomfortably large for a buffer that will be handled as an atomic entity for network purposes - and you're also going to be sacrificing latency for other requests in the same h2 connection. It would be interesting (but definitely out of scope for this) to figure out the size at which the returns in throughput start diminishing: I can believe that the fragmentation at 16KiB makes the overhead noticeable, but is that still the case at 64? at 256?

Reply:
Yeah, I agree with you there. This topic interests me a lot - so just for fun, some graphs! Each run is 60 seconds of a stream (with 1 client) sending messages as fast as it can, and the server echoing the response back. Would need way more data to get any real conclusions of course, but looks like maybe the best size could lie somewhere between 512KiB and 1 MiB. 1MiB is about where things drop off, the difference between 1 and 4 MiB is pretty small too. 


![Figure_1](https://github.com/kralicky/jobserver/assets/8194899/aa2f6d91-0922-4d05-9a36-b213c636847a)
![Figure_2](https://github.com/kralicky/jobserver/assets/8194899/168a182b-4084-4ca4-abf8-e6d0bdb73b9c)

| Payload Size (KiB) | Mbps | msgs/s |
| --- | --- | --- |
| 16 | 580.7 | 278731 |
| 24 | 672.7 | 215249 |
| 32 | 844 | 202569 |
| 48 | 901.1 | 144174 |
| 64 | 996.9 | 119625 |
| 96 | 1266.2 | 101294 |
| 128 | 1300.5 | 78030 |
| 192 | 1431.7 | 57267 |
| 256 | 1665.3 | 49959 |
| 384 | 1782.5 | 35650 |
| 512 | 1968.7 | 29530 |
| 768 | 2373 | 23730 |
| 1024 | 2364.1 | 17731 |
| 2048 | 2582.4 | 9684 |
| 4096 | 2686.9 | 5038 |
| 5120 | 2723.3 | 4085 |
| 6144 | 2771.2 | 3464 |
| 7168 | 2685.2 | 2877 |
| 8192 | 2621.9 | 2458 |
| 16384 | 2596.3 | 1217 |
| 24576 | 2556.8 | 799 |
| 32768 | 2781.9 | 652 |
| 49152 | 2796.8 | 437 |
| 65536 | 3029.3 | 355 |